### PR TITLE
API Consumers & Providers

### PR DESCRIPTION
**Pull request by Fox32**
This displays both the consumers and providers in the API page and also the consumed APIs in the component page:

Bildschirmfoto 2020-11-27 um 17 07 50
Bildschirmfoto 2020-11-27 um 17 06 59

image

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)

### CODE REVIEW

**Comment by freben** [comment]
Woo! I love seeing the relations starting to get some usage.

Not reviewing right now, but one unrelated thought is whether empty states in cards should have a collapsed form that's shown by default. Like,

`No consumers.      [Show Details]`

And then when you press the details button we could either expand the contents into you have in the screenshot, or open a popover with all the nitty gritty details

**Comment by stefanalund** [comment]
This is truly great. I would love to see if/how dependency graphs could be used to improve the experience. As you say, that will be nicer than tables.

**Comment by Fox32** [comment]
Made some experiments with a relations graph. That could be included as a widget in the catalog.

image

I'm not sure yet if it's really useful, but it looks cool 😉 The nodes could be full cars, including the owner and/or some more details.

What do you think? I could include both variants in this PR, or just start with the tables and do another PR for the graph.

**Comment by andrewthauer** [comment]
I think the dependency graph could be a nice alternative view + demonstrate the components usage. I'm curious if the plumbing is in place to easily get dependencies of dependencies. This is where graph could start becoming an interesting view imo.

**Comment by Fox32** [comment]
> I think the dependency graph could be a nice alternative view + demonstrate the components usage. I'm curious if the plumbing is in place to easily get dependencies of dependencies. This is where graph could start becoming an interesting view imo.

Sure, I could add some kind of `depth` parameter and fetch multiple levels. I will take a look!

**Comment by freben** [comment]
Yeah it'd take multiple requests at the moment. But that's probably OK, and may be easier when the graphql layer is in place so you can do `deps { deps { deps ...` at least down to some depth :)

**Comment by Fox32** [comment]
> Yeah it'd take multiple requests at the moment. But that's probably OK, and may be easier when the graphql layer is in place so you can do `deps { deps { deps ...` at least down to some depth :) Or a tree-as-table `recursiveDeps` thing of course.

Yes, graphql would be useful here. The example above uses multiple requests, I would stay with this for now.

**Comment by Fox32** [comment]
I continued to play around with it. Having all relationships included might now be useful. This way I have a lot of duplicates:
image

It would be better if I only include the relationships from the direction of the start entity.

image


As this PR is becoming quite huge, I would be up to moving the last commit into a separate PR and start only with the tables. I can also move the new tests into a separate PR, I added them because we miss coverage, but it's not required here. What do you think?

**Comment by andrewthauer** [comment]
@Fox32 - I'd argue that it would be useful to see what are essentially circular references. However, maybe a toggle. But agree, this could be iterated on. Great work on the experiment though 👍. This excites me with the possibilities it is starting to show.

**Comment by freben** [comment]
Yes agreed for sure - limit scope and break into several PRs!

Personally, I'd be up for the graph thing to be on a completely separate page (or as a popup dialog or something). In my mind, when the relations grow large, the most useful data display would be a list of tables, one per relation.

```
Owned By
    [ freben | user | bleh | blah ]
Maintained By
    [ admin-1 | user | Some Full Name | blah ]
    [ devops | group | The Devops | blah ]
Consumes APIS
    [ X | column | blah ]
    [ Y | column | blah ]
```

etc. Clickable of course, and perhaps with topical icons

**Comment by Fox32** [comment]
Move some changes to #3506 and #3508

File: packages/catalog-model/examples/all-apis.yaml [hunk 1]
```diff
@@ -11,3 +11,5 @@ spec:
     - https://github.com/backstage/backstage/blob/master/packages/catalog-model/examples/apis/spotify-api.yaml
     - https://github.com/backstage/backstage/blob/master/packages/catalog-model/examples/apis/streetlights-api.yaml
     - https://github.com/backstage/backstage/blob/master/packages/catalog-model/examples/apis/swapi-graphql.yaml
+    - https://github.com/backstage/backstage/blob/master/packages/catalog-model/examples/apis/wayback-archive-api.yaml
+    - https://github.com/backstage/backstage/blob/master/packages/catalog-model/examples/apis/wayback-search-api.yaml
```
Comment (Fox32) [inline packages/catalog-model/examples/all-apis.yaml:15]: This is something I have to change after #3513

File: packages/catalog-model/examples/all-components.yaml [hunk 2]
```diff
@@ -15,3 +15,5 @@ spec:
     - https://github.com/backstage/backstage/blob/master/packages/catalog-model/examples/components/playback-lib-component.yaml
     - https://github.com/backstage/backstage/blob/master/packages/catalog-model/examples/components/www-artist-component.yaml
     - https://github.com/backstage/backstage/blob/master/packages/catalog-model/examples/components/shuffle-api-component.yaml
+    - https://github.com/backstage/backstage/blob/master/packages/catalog-model/examples/components/wayback-archive-component.yaml
+    - https://github.com/backstage/backstage/blob/master/packages/catalog-model/examples/components/wayback-search-component.yaml
```
Comment (Fox32) [inline packages/catalog-model/examples/all-components.yaml:19]: Here too

**Comment by Fox32** [comment]
So having the two PR above merged, code size is now a lot smaller 🙂 

I also removed the relations component for now because of the mixed feedback. If you are interested, I can add it in a later PR. For now I keep an backup of the changes here: https://github.com/backstage/backstage/compare/master...SDA-SE:feat/api-docs-relation-card?expand=1

**Review by Rugvip — COMMENTED** [review]
Looks good, thanks! :grin:

Could you add instructions for how to update the app? similar to existing examples in https://github.com/backstage/backstage/blob/master/packages/create-app/CHANGELOG.md. Otherwise good to go imo :shipit:

**Comment by Fox32** [comment]
@Rugvip gave this an update as discussed

File: plugins/api-docs/src/components/useRelatedEntities.ts [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { Entity } from '@backstage/catalog-model';
+import { useApi } from '@backstage/core';
+import { catalogApiRef } from '@backstage/plugin-catalog';
+import { useAsyncRetry } from 'react-use';
+
+// TODO: Maybe this hook is interesting for others too?
+export function useRelatedEntities(
```
Comment (Rugvip) [inline plugins/api-docs/src/components/useRelatedEntities.ts:22]: Another one for the discussion in #3623 x)

**Review by Rugvip — APPROVED** [review]
Awesome, thanks! 😁

**Pull request merged**